File: java/collatz.java
import java.util.*;

// Object oriented design is best suited for larger applications.
// To take advantage of this, Object-oriented languages require double the amount of code.

class Collatz
{
	public static void main(String args[]){
		//setup
		long i = 1;
		long length = 0;
		long tenth = 0;
		ArrayList<LongPair> biggest = new ArrayList();
		for(i = 0; i < 10; i++){
			biggest.add(new LongPair());
		}
		
		//calculation
		for(long n = 2; n <= 5000000L; n++){
			i = n;
			length = 0;
			while(i >1){
				if(i % 2 == 1){
					i = 3*i+1;
				}else{
					i = i >> 1;
				}
				length++;
			}
			//longest element tracking
			if(length > tenth){
				//System.out.println(n + " -> " + length + ", \t " + tenth);
				biggest.add(new LongPair(length, n));
				Collections.sort(biggest, new LPcompar());
				biggest.remove(0);
				tenth = biggest.get(0).first;
			}
		}
		//display
		System.out.println("Largest elements:\n\nBy length:");
		for(LongPair x : biggest){
			System.out.println("Number = " + x.second + ", Length = " + x.first);
			x.swap();
		}
		System.out.println("\nBy size:");
		Collections.sort(biggest, new LPcompar());
		for(LongPair x : biggest){
			System.out.println("Number = " + x.first + ", Length = " + x.second);
		}
	}
}

//Java doesn't seem to have a standard pair class, so I'll write one up.
class LongPair{
	public long first;
	public long second;
	
	//constructors
	LongPair(){
		first = 0;
		second = 0;
	}
	LongPair(long a, long b){
		first = a;
		second = b;
	}
	
	//swap the 2 elements
	public void swap(){
		long temp = first;
		first = second;
		second = temp;
	}
}

//To use Collections.sort(), I need to write a separate comparator class for my pair.
class LPcompar implements Comparator<LongPair>{
	public int compare(LongPair left, LongPair right){
		return (int)(left.first - right.first);
	}
}

File: c#/collatz.cs
using System;
using System.Collections.Generic;

class Collatz{
	static void Main(string[] args){
		ulong length = 0;
		ulong tenth = 0;
		//c# does not have typedefs, which I would like.
		//at least it has a pair class.
		List<Tuple<ulong, ulong> > biggest = new List<Tuple<ulong, ulong> >();
		ulong i = 0;
		for(i = 0; i < 10; i++){
			biggest.Add(new Tuple<ulong, ulong>(0, 0));
		}
		//calculation
		for(ulong n = 2; n <= 10000000; n++){
			i = n;
			length = 0;
			while(i > 1){
				if((i & 1) != 0){
					i = i*3+1;
				}else{
					i = i >> 1;
				}
				length++;
			}
			//largest element tracking
			if(length >= tenth){
				//Console.WriteLine("{0} -> {1}", n, length);
				biggest.RemoveAt(0);
				//COOL THING: C# compares tuples by their first element by default.
				biggest.Add(new Tuple<ulong, ulong>(length, n));
				//sort
				biggest.Sort();
				tenth = biggest[0].Item1;
			}
		}
		Console.WriteLine("Top 10 largest elements:\n\nBy Length:");
		for(int j = 0; j < 10; j++){
			//C#'s string formatting is an interesting alternative to the common prinf format.
			Console.WriteLine("Number: {0}, \tLength: {1}",
				biggest[j].Item2.ToString(), biggest[j].Item1.ToString());
			//swap
			biggest[j] = new Tuple<ulong, ulong>(biggest[j].Item2, biggest[j].Item1);
		}
		biggest.Sort();
		Console.WriteLine("\nBy Size:");
		for(int j = 0; j < 10; j++){
			Console.WriteLine("Number: {0}, \tLength: {1}",
				biggest[j].Item1.ToString(), biggest[j].Item2.ToString());
		}
	}
}
